### ISSUE TITLE
compaction: Don't split user keys during compactions

### ISSUE THREAD

**Issue by itsbilal** [cockroachdb/pebble#734]
Currently, there are some cases in [the compaction loop](https://github.com/cockroachdb/pebble/blob/master/compaction.go#L1516) where user keys could be output to two different sstables, like this:

```
000157.sst:
a.RANGEDEL.6:f
b.SET.15:foo
b.MERGE.8:bar

000158.sst:
b.RANGEDEL.6:f
b.SET.5:baz
b.DEL.3

```

This results in the implicit creation of an "atomic compaction group"; both these SSTables must
be present in compactions together, or it's possible for deleted keys to reappear after a sequence of compactions (see [the comment above](https://github.com/cockroachdb/pebble/blob/master/compaction.go#L306) `expandInputs` on how and why this happens).

These implications of splitting user keys are ultimately unhelpful in increasing compaction parallelization and reducing compaction sizes. The only reason why we split user keys is to maintain similarity in behaviour with RocksDB. We should explore not splitting user keys across different sstables for all compactions (something we already do for flushes as of #675). This should help simplify some compaction logic.

**Comment by petermattis** [comment]
> I thought the RocksDB motivation was to prevent a single sstable from getting too large if there were too many seqnums for a key, which can result in larger number of index blocks (2-level indexes) and bigger range tombstone block which can affect read performance.

The sstable splitting logic in RocksDB was inherited from LevelDB which didn't need to worry about atomic compaction units because it doesn't have range tombstones. I looked at the history of how this was added, and my read was that splitting at the next user key was simply not considered as an alternative to atomic compaction units. I'm not sure why.

How much bigger can the range tombstone block get from splitting at the next user key? I suspect very little in practice, but perhaps I'm not imagining some problematic corner case.

How much bigger can the index block get from splitting at the next user key? I think that is dependent on how many snapshots are active. AFAIK, we only use snapshots for replica rebalancing, so we're going to only have a few active at any time. Even if there were 1000 snapshots active, that translates to the potential to add 1000 extra records to the sstable which seems reasonable for most scenarios in which L6 sstables typically have hundreds of thousands of records.
